tests/virt_autotest/virsh_virtual_network.pm

```diff
@@ -0,0 +1,126 @@
+	assert_script_run("ssh root\@$guest 'cd /etc/sysconfig/network/; cp ifcfg-eth0 ifcfg-eth1'", 60);
+	record_info "virsh-virtual_network", "NAT BASED NETWORK";
+	assert_script_run("echo '$vnet_nated' >> vnet_nated");
+	assert_script_run("virsh net-create vnet_nated");
+	assert_script_run("virsh net-info vnet_nated");
+	assert_script_run("virsh attach-interface $guest network vnet_nated --model virtio --live");
+	my $gi_vnet_nated = script_output("journalctl | grep dnsmasq-dhcp| tail -1| grep -Po "\d+\.\d+\.\d+\.\d+"");
+	assert_script_run("ssh $gi_vnet_nated ip route");
+	save_screenshot;
+	assert_script_run("virsh detach-interface $guest network --current");
+	assert_script_run("virsh net-destroy vnet_nated");
+	record_info "virsh-virtual_network", "BRIDGE - USING AN EXISTING BRIDGE ON VM HOST SERVER";
+	assert_script_run("echo '$vnet_host_bridge' >> vnet_host_bridge");
+        assert_script_run("virsh net-create vnet_host_bridge");
+        assert_script_run("virsh attach-interface $guest network vnet_host_bridge --model virtio --live");
+        my $mac_vnet_host_bridge = script_output("virsh domiflist $guest | grep vnet_host-bridge| grep -E -o '[[:xdigit:]]{2}(:[[:xdigit:]]{2}){5}'");
+        my $gi_vnet_host_bridge = script_output("arp | grep $mac_vnet_isolated_bridge| awk \'{print $1}\'");
+	assert_script_run("ssh $gi_vnet_host_bridge ip route");
+	save_screenshot;
+	assert_script_run("virsh detach-interface $guest bridge --mac $my_host_br_mac --current");
+	assert_script_run("virsh net-destroy vnet_host_bridge");
+        record_info "virsh-virtual_network", "BRIDGE - ISOLATED NETWORK";
+	assert_script_run("echo '$vnet_isolated_bridge' >> vnet_isolated_bridge");
+        assert_script_run("virsh net-create vnet_isolated_bridge");
+        assert_script_run("virsh attach-interface $guest network vnet_isolated_bridg --model virtio --live");
+	my $gi_vnet_isolated_bridge = script_output("journalctl | grep nsmasq-dhcp| tail -1| grep -oE '[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}'");
+	assert_script_run("ssh $gi_vnet_isolated_bridg ip route");
+	save_screenshot;
+	assert_script_run("virsh detach-interface $guest network --current");
+	assert_script_run("virsh net-destroy vnet_isolated_bridge");
+	record_info "virsh-virtual_network", "ROUTED NETWORK";
+	assert_script_run("echo '$vnet_routed' >> vnet_routed");
+	assert_script_run("virsh net-define vnet_routed");
+	assert_script_run("virsh net-start vnet_routed");
+	assert_script_run("virsh attach-interface $guest network vnet_routed --model virtio --live");
+	my $gi_vnet_routed = script_output("journalctl | grep dnsmasq-dhcp| tail -1| grep -oE "([0-9]{1,3}[\.]){3}[0-9]{1,3}"");
+	assert_script_run("ssh $gi_vnet_isolated_bridg ip route");
+	save_screenshot;
+	assert_script_run("virsh detach-interface $guest network --current");
+	assert_script_run("virsh net-destroy vnet_routed");
+	assert_script_run("virsh virsh net-undefine vnet_routed");
+
+    }
```

Comment:
I prefer group test steps by types of network, for example, 
    record_info "FOR NETWORK XXX";
    foreach my $guest (keys %xen::guests) {
    }
    record_info "FOR NETWORK YYY";
    foreach my $guest (keys %xen::guests) {
    }

Reply:
Awesome, great idea. I am working on it as your mentioned.